### PR TITLE
feat: create signaling bridge

### PR DESCRIPTION
**Pull request by rodrigoleonello**

### CODE REVIEW

File: src/Main.cpp [hunk 1]
```diff
@@ -1,10 +1,467 @@
...
+                else if (actiontype == "candidate")
+                {
+                    candidateRustyMessageParser(signalr_websocket, rusty_decoder);
+                }
+            }
+        }
+    );
+
+    // wss://signalserverhost?user=yourname
+    const string rusty_url_websocket = "127.0.0.1:3012?user=" + local_peer_id;
```
Comment (doudou) [inline src/Main.cpp:251]: Same. This needs to be configurable. Maybe actually make them configurable right now ? As arguments to the binary...

File: src/Main.cpp [hunk 2]
```diff
@@ -1,10 +1,467 @@
...
+        Json::FastWriter fast;
+        ws->send(fast.write(message));
+    }
+}
+
+bool initialHandShakeFinalized(
+    shared_ptr<rtc::WebSocket>& websocket,
+    SignalRMessageDecoder& decoder,
+    string& local_peer_id,
+    bool& client_id_checked
```
Comment (doudou) [inline src/Main.cpp:447]: Right

So, this function returns true when the client id is checked ... Just use the return value to set `client_id_checked` in the caller.

Reminder: https://docs.google.com/document/d/16zBGRa58zu7wgRBcGb6IVVQtCfiDqDDhu4vjUVnVSu0/edit#heading=h.tu1h6jll3wc

**Review by jhonasiv — APPROVED** [review]

File: src/Main.cpp [hunk 3]
```diff
@@ -1,10 +1,466 @@
...
+    {
+        throw invalid_argument(error);
+    }
+    string https_token = decoder.getConnectionToken();
+
+    SignalRMessageDecoder signalr_decoder = SignalRMessageDecoder();
+    promise<void> wsr_promise;
+    future<void> wsr_future = wsr_promise.get_future();
+    rtc::Configuration signalr_config;
+    signalr_config.iceServers.emplace_back("stun:" + stun_server);
```
Comment (jhonasiv) [inline src/Main.cpp:101]: Clean some of this up, it is a bit hard to read. Split some commands in groups separated by spaces.

File: src/Main.cpp [hunk 4]
```diff
@@ -1,10 +1,466 @@
...
+    {
+        Json::FastWriter fast;
+        ws->send(fast.write(message));
+    }
+}
+
+void sendSessionCheck(shared_ptr<rtc::WebSocket>& websocket, string& local_peer_id)
+{
+    Json::Value message, content(Json::arrayValue);
+    content.append(local_peer_id);
```
Comment (jhonasiv) [inline src/Main.cpp:302]: Isnt this supposed to be the client_id? Also, I couldnt find out how you would get the client_id by now.

File: src/Main.cpp [hunk 5]
```diff
@@ -1,10 +1,466 @@
...
+    SignalRMessageDecoder& decoder,
+    string& local_peer_id
+)
+{
+    if (decoder.isEmpty())
+    {
+        sendSessionCheck(websocket, local_peer_id);
+        return false;
+    }
+    else if (decoder.checkSessionList())
```
Comment (jhonasiv) [inline src/Main.cpp:455]: I would imagine you would get the client id from the session list, maybe it isnt really needed to perform the session check?

File: src/Main.cpp [hunk 6]
```diff
@@ -1,10 +1,466 @@
...
+
+void joinSession(
+    shared_ptr<rtc::WebSocket>& websocket,
+    SignalRMessageDecoder& decoder,
+    string& local_peer_id
+)
+{
+    Json::Value message, content(Json::arrayValue);
+    content.append(local_peer_id);
+    content.append(decoder.getSessionIdFromList());
```
Comment (jhonasiv) [inline src/Main.cpp:323]: Not sure what type content is here, but it is supposed to be an array of dumped JSONs in the following form:
```
arguments: [
    "{client_id: \"the id here\", session_id: \"the session id here\"}
]
```

Probably with the keys transformed into string forms too? Maybe deep trekker didnt notice this when they wrote their docs.

File: src/HttpPostRequestReplyMessageDecoder.cpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    if (!value.isMember(fieldName))
+    {
+        throw invalid_argument("message does not contain the " + fieldName + " field");
+    }
+}
+
+string HttpPostRequestReplyMessageDecoder::getConnectionToken()
+{
+    validateFieldPresent(mJData, "connectionToken");
+    return mJData["action"].asString();
```
[the remark this answers is not in the document]
Comment (rodrigoleonello) [reply src/HttpPostRequestReplyMessageDecoder.cpp:28]: yes.
Copy/paste error...

File: src/Main.cpp (hunk 4 above)
Comment (rodrigoleonello) [reply src/Main.cpp:302]: The `local_peer_id` is the `client_id`, from the deep_trekker perspective. They are the same.

File: src/Main.cpp (hunk 5 above)
Comment (rodrigoleonello) [reply src/Main.cpp:455]: The check is basically to confirm that the deep_trekker got our `local_peer_id` as `client_id`. So we should get the id from the list and compare with the local_peer_id.

File: src/Main.cpp (hunk 6 above)
Comment (rodrigoleonello) [reply src/Main.cpp:323]: Actually the print, in the website, confirm what you sad. I'm not parsing correctly the msg.

Comment (rodrigoleonello) [reply src/Main.cpp:323]: Unless in the session_check step (topic 5), that there is no key for the client_id.

**Pull request merged**